### ISSUE TITLE
Spurious "function must be explicitly imported to be extended" error

### ISSUE THREAD

**Issue by tkelman** [JuliaLang/julia#15455]
Reduced from https://github.com/JuliaLang/WinRPM.jl/pull/73 - `select` is not being exported from this module, so it shouldn't need to import `Base.Sort.select` to create its own private unexported `select` method, right?
```jl
module minrpm

immutable ParsedData
end

immutable Packages{T<:Union{Set{ParsedData},Vector{ParsedData}}}
    p::T
end

function rpm_provides{T<:AbstractString}(requires::Union{Vector{T},Set{T}})
    pkgs = Set{ParsedData}()
    for x in requires
        pkgs_ = rpm_provides(x)
        if isempty(pkgs_)
            warn("Package not found that provides $x")
        else
            push!(pkgs, select(pkgs_,x))
        end
    end
    Packages(pkgs)
end

function select(pkgs::Packages, pkg::AbstractString)
    pkg
end

end
```

At dbfc0e2 (and latest master) this gives
```
 _/ |\__'_|_|_|\__'_|  |  Commit dbfc0e2* (25 days old master)
|__/                   |  i686-w64-mingw32

julia> include("minrpm.jl")
ERROR: LoadError: error in method definition: function Sort.select must be explicitly imported to be extended
 in include(::ASCIIString) at .\boot.jl:264
 in include_from_node1(::ASCIIString) at .\loading.jl:417
 in eval(::Module, ::Any) at .\boot.jl:267
while loading C:\cygwin64\home\Tony\julia32\minrpm.jl, in expression starting on line 23
```

One commit earlier, at e0bc056 it imports without error
```
 _/ |\__'_|_|_|\__'_|  |  Commit e0bc056* (25 days old master)
|__/                   |  i686-w64-mingw32

julia> include("minrpm.jl")
minrpm

julia> using minrpm
```
cc @vtjnash, ref https://github.com/JuliaLang/julia/pull/15064

**Comment by tkelman** [comment]
Symptom is the same as #15264, but was not fixed by #15335

**Comment by vtjnash** [comment]
it looks like the same issue as #15264 (the usage of `select` in the first function body causes its binding to be resolved before `select` is defined)
